### PR TITLE
ENH: beta negative binomial distribution

### LINKED ISSUE #17330: ENH: Implement Beta negative binomial distribution
### Is your feature request related to a problem? Please describe.

This is a distribution I would find useful in the library https://en.wikipedia.org/wiki/Beta_negative_binomial_distribution 
I would like to be able to calculate the probability mass function etc. like with other distributions.

### Describe the solution you'd like.

_No response_

### Describe alternatives you've considered.

_No response_

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Closes https://github.com/scipy/scipy/issues/17330

#### What does this implement/fix?
Adds the beta negative binomial distribution as `stats.betanbinom`.

#### Additional information
Pinging @jachymb who issued #17330: any help with the review is much appreciated.

The formulas for the moments were taken from [Wolfram Alpha](https://www.wolframalpha.com/input?i=BetaNegativeBinomialDistribution%5B%5C%5BAlpha%5D%2C+%5C%5BBeta%5D%2C+n%5D).

Wolfram Alpha also contains a function for the CDF:
image

I decided not to implement that here as I am not sure if a naive implementation using the `special` functions (especially [`special.hyperu`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.hyperu.html#scipy.special.hyperu), if that is the same generalized hypergeometric function from the Formula) will push this PR in a potentially very deep rabbit hole.

Somehow the `expect` method does not work with this distribution. I did not dig deeper in the discrete distribution infrastructure there though as that will be overhauled in future from what I understood.

Example:

```python

n, a, b = 5, 4.3, 1
from scipy.stats import betanbinom

dist = betanbinom(n, a, b)
dist.stats(moment="mvsk")

#  (1.5151515151515151, 7.1246057412065324, 477.61561430806694, 502.0609267840596)

dist.expect(lambda x: x), dist.expect(lambda x: x**2)
# RuntimeWarning: expect(): sum did not converge
#  warnings.warn('expect(): sum did not converge', RuntimeWarning)
# (1.5151510425851544, 9.419593400972543)
```

The results are similarly off with different parameter combinations for `n, a, b`.

### CODE REVIEW

**Comment by chrisb83** [comment]
Hi Daniel,

I reviewed the PR. Documentation / tutorial are fine. The implementation of `pdf` (and the resulting `cdf`) look good, I ran a few more test cases (I will add the code below).

The only concern I have is skew/kurtosis. I noted that you skip `check_kurt_expect`. This test compares the following quantities:
the central fourth moment using `expect` and `(k+3)*v**2`. `expect` might not be accurate, but replacing it with its empirical estimate, I also cannot reconcile the values. Here is my code: 

```
from scipy import stats
import numpy as np

dist = stats.betanbinom(5, 4.3, 1)
r = dist.rvs(size=1_000_000)
m, v, s, k = dist.stats('mvsk')

# empirical estimates
me, ve = np.mean(r), np.var(r, ddof=1)
se = stats.skew(r)
ke = stats.kurtosis(r)

print(f"m: {m}, {me}")
print(f"v: {v}, {ve}")
print(f"s: {s}, {se}")
print(f"k: {k}, {ke}")

# comparison as in `check_kurt_expect`
m4e_est = np.mean((r - m)**4)
print((ke + 3)*ve**2, m4e_est, (k+3)*v**2)
```

Output:
```
m: 1.5151515151515151, 1.514323
v: 7.1246057412065324, 7.07704592871693
s: 477.61561430806694, 5.976190300924765
k: 502.0609267840596, 128.84108262027212
6603.205128114574 6602.8190767446085 25636.896162638084
```

Could you check please? We should add a separate test for kurtosis.

**Comment by chrisb83** [comment]
Test cases for pdf and cdf:

```
import pytest
from numpy.testing import assert_allclose
from mpmath import mp
from scipy.stats import betanbinom


mp.dps = 100

# PDF
def betanbinom_pmf(k, n, a, b):
    k = mp.mpf(k)
    a = mp.mpf(a)
    b = mp.mpf(b)
    n = mp.mpf(n)
    return float(mp.binomial(n + k - mp.one, k) * mp.beta(a + n, b + k)
                    / mp.beta(a, b))

# CDF: just sum up the pdf values computed for each value j=0,...,k
def betanbinom_cdf(k, n, a, b):
    res = 0
    a = mp.mpf(a)
    b = mp.mpf(b)
    n = mp.mpf(n)

    denom = mp.beta(a, b)
    for j in mp.arange(k+1):
        p = mp.binomial(n + j - mp.one, j) * mp.beta(a + n, b + j)
        res += p
    return float(res / denom)

test_vals_pdf = []
beta_params = [0.05, 0.5, 1, 5, 7, 20, 100]
for a, b in zip(beta_params, beta_params):
    for n in [5, 10, 100, 1000, 10000]:
        for x in [1, 5, 100, 1000, 10000, 100000]:
            expected = betanbinom_pmf(x, n, a, b)
            test_vals_pdf.append([x, n, a, b, expected])


@pytest.mark.parametrize('x, n, a, b, ref', test_vals_pdf)
def test_betanbinom_pmf(x, n, a, b, ref):
    assert_allclose(betanbinom.pmf(x, n, a, b), ref, rtol=1e-10)

beta_params = [0.05, 0.5, 5, 20]
test_vals_cdf = []
for a, b in zip(beta_params, beta_params):
        for n in [5, 10, 100, 1000]:
            for x in [1, 5, 100, 1000, 10000]:
                expected = betanbinom_cdf(x, n, a, b)
                #print([x, n, a, b, expected])
                test_vals_cdf.append([x, n, a, b, expected])

@pytest.mark.parametrize('x, n, a, b, ref', test_vals_cdf)
def test_betanbinom_cdf(x, n, a, b, ref):
    assert_allclose(betanbinom.cdf(x, n, a, b), ref, rtol=1e-10)

```
Output: 255 passed and no failures

The test for the cdf is pretty slow (the above code runs a bit more than a minute, that is why I tested less values than for the cdf). If one adds `x=100000` to the list used in the pdf test, one gets a few (20 out of 210) failures with tol=1e-10, but none with tol=1e-9. Those failures are all for very extreme values, similar to the following example:

```
x = 100000, n = 1000, a = 100, b = 100, ref = 6.447038317256134e-144
>       assert_allclose(betanbinom.pmf(x, n, a, b), ref, rtol=1e-10)
E       AssertionError:
E       Not equal to tolerance rtol=1e-10, atol=0
E       Mismatched elements: 1 / 1 (100%)
E       Max absolute difference: 1.39764453e-153
E       Max relative difference: 2.16788618e-10
E        x: array(6.447038e-144)
E        y: array(6.447038e-144)
```

**Comment by dschmitz89** [comment]
> Hi Daniel,
> 
> I reviewed the PR. Documentation / tutorial are fine. The implementation of `pdf` (and the resulting `cdf`) look good, I ran a few more test cases (I will add the code below).
> 
> The only concern I have is skew/kurtosis. I noted that you skip `check_kurt_expect`. This test compares the following quantities: the central fourth moment using `expect` and `(k+3)*v**2`. `expect` might not be accurate, but replacing it with its empirical estimate, I also cannot reconcile the values. Here is my code:
> 
> ```
> from scipy import stats
> import numpy as np
> 
> dist = stats.betanbinom(5, 4.3, 1)
> r = dist.rvs(size=1_000_000)
> m, v, s, k = dist.stats('mvsk')
> 
> # empirical estimates
> me, ve = np.mean(r), np.var(r, ddof=1)
> se = stats.skew(r)
> ke = stats.kurtosis(r)
> 
> print(f"m: {m}, {me}")
> print(f"v: {v}, {ve}")
> print(f"s: {s}, {se}")
> print(f"k: {k}, {ke}")
> 
> # comparison as in `check_kurt_expect`
> m4e_est = np.mean((r - m)**4)
> print((ke + 3)*ve**2, m4e_est, (k+3)*v**2)
> ```
> 
> Output:
> 
> ```
> m: 1.5151515151515151, 1.514323
> v: 7.1246057412065324, 7.07704592871693
> s: 477.61561430806694, 5.976190300924765
> k: 502.0609267840596, 128.84108262027212
> 6603.205128114574 6602.8190767446085 25636.896162638084
> ```
> 
> Could you check please? We should add a separate test for kurtosis.

Thanks for the review @chrisb83 .

I corrected two mistakes in the skewness and kurtosis formulas. The skewness looks good now, kurtosis I am still not 100% sure though.

```python

from scipy import stats
import numpy as np

dist = stats.betanbinom(10, 8, 6)
r = dist.rvs(size=1_000_000)
m, v, s, k = dist.stats('mvsk')

# empirical estimates
me, ve = np.mean(r), np.var(r, ddof=1)
se = stats.skew(r)
ke = stats.kurtosis(r)

print(f"s: {s}, {se}")

# 2.1824847809330823, 2.1734327575959322

print(f"k: {k}, {ke}")

# 13.411674208144797, 10.286132293356905
```

With dedicated test you mean a dedicated test for the kurtosis in `test_discrete_distns`?

**Comment by chrisb83** [comment]
> With dedicated test you mean a dedicated test for the kurtosis in `test_discrete_distns`?

yes, that is what I meant. Essentially you can use something similar like: `m4e_est = np.mean((r - m)**4)` and then `m4e_est` and `(k+3)*v**2` should be closed (notation as above)

I would also say that kurtosis value should be closer to the empirical estimate. From where did you take the expression for kurtosis?

**Comment by dschmitz89** [comment]
The formula is from Wolfram: https://www.wolframalpha.com/input?i=BetaNegativeBinomialDistribution%5B%5C%5BAlpha%5D%2C+%5C%5BBeta%5D%2C+n%5D

But I am starting to doubt it's correctness.

**Comment by josef-pkt** [comment]
> print(f"k: {k}, {ke}")
# 13.411674208144797, 10.286132293356905

if the difference is 3, then it is most likely pearson versus fisher kurtosis

> ke = stats.kurtosis(r)

scipy defaults to  fisher kurtosis which subtracts 3

**Comment by mdhaber** [comment]
Sometimes empirical estimates don't converge for any practical number of observations. I'd suggest using `mpmath.nsum`.

**Comment by chrisb83** [comment]
Have you checked a few more cases if the issue was indeed fisher vs pearson kurtosis? I tried a few parameters and the differences between k and ke were often very large. As Matt said, it could be an issue that the empirical estimates are not accurate, so I am not saying that the formula is wrong. I just think we should look at a few more examples and add a test case.

**Comment by dschmitz89** [comment]
> Have you checked a few more cases if the issue was indeed fisher vs pearson kurtosis? I tried a few parameters and the differences between k and ke were often very large. As Matt said, it could be an issue that the empirical estimates are not accurate, so I am not saying that the formula is wrong. I just think we should look at a few more examples and add a test case.

In the beginning, there was a typo in the formula which I corrected in bd52dc201bea821d3b7fee01295d43524428903c
. The correct formula and accounting for the different definitions results yields marching results now. I will post some more thorough checks in the next days.

File: scipy/stats/_discrete_distns.py
Comment (josef-pkt) [inline scipy/stats/_discrete_distns.py:463]: I think this is wrong.
AFAIR, all distributions return pearson kurtosis. Only the stats function kurtosis defaults to excess kurtosis

**Comment by josef-pkt** [comment]
One problem for checking and unit test is that for large kurtosis we need huge samples (like the 1 million).
Estimates for higher moments in heavy tailed distributions are very noisy (have huge variance).

(AFAIR, variance of kurtosis estimate depends on 8th moment, so alpha would have to be larger than 8 to have finite variance of the kurtosis estimate.)

**Comment by dschmitz89** [comment]
I tested out a wider range of parameters and the moments matched. I will take a few extreme parameters as unit test.

```python

from scipy import stats
import numpy as np
from numpy.testing import assert_allclose

rng = np.random.default_rng()

n_samples = rng.integers(0, 1000, size=10)
a_samples = rng.uniform(8, 100, size=10)
b_samples = rng.uniform(1, 100, size=10)
print(n_samples)
print(a_samples)
print(b_samples)

for n in n_samples:
    for a in a_samples:
        for b in b_samples:
            dist = stats.betanbinom(n, a, b)

            # analytical moments
            m, v, s, k = dist.stats('mvsk')
            moments_analytic = [m, v, s, k]

            # empirical estimates
            r = dist.rvs(size=5000000)
            me = np.mean(r)
            ve = np.var(r, ddof=1)
            se = stats.skew(r)
            ke = stats.kurtosis(r)
            moments_empirical = [me, ve, se, ke]
            try:
                assert_allclose(moments_analytic, moments_empirical, rtol=5e-2)
            except AssertionError:
                print("========ERROR===========")
                print(f"n: {n}, a: {a}, b: {b}")
                print(f"m: {m}, me: {me}")
                print(f"v: {v}, ve: {ve}")
                print(f"s: {s}, se: {se}")
                print(f"k: {k}, ke: {ke}")

```

File: scipy/stats/_discrete_distns.py
Comment (chrisb83) [reply scipy/stats/_discrete_distns.py:463]: It should be ok: `stats.norm().stats('k')` returns 0 (Fisher kurtosis)

File: scipy/stats/_distr_params.py
Comment (chrisb83) [inline scipy/stats/_distr_params.py:128]: I recommend changing the default parameters. Hopefully the default kurtosis test passes in that case. With a=4.3, the kurtosis estimate is well-defined but its estimate is not stable, I assume that also explains why expect has problems. Maybe take (5, 9.3, 1)

File: scipy/stats/tests/test_discrete_basic.py
Comment (chrisb83) [inline scipy/stats/tests/test_discrete_basic.py:95]: see above, this is not needed if we change the default parameters. I tried to run the test with 5, 9.3, 1 and it was ok

File: scipy/stats/tests/test_discrete_distns.py
Comment (chrisb83) [inline scipy/stats/tests/test_discrete_distns.py:619]: k = betanbinom.stats(n, a, b, moments="k")

add a blank line at the end of the file

**Review by chrisb83 — COMMENTED** [review]
I will approve this PR once a few minor points are addressed

**Comment by chrisb83** [comment]
> I tested out a wider range of parameters and the analytical and empirical moments matched. I will take a few extreme parameters as unit test. Increasing the number of samples to 5 mio. and using 
 was crucial to get the accuracy.

yes, I agree that it should be ok now. For a < 8, probably the point Josef made about the variance of the estimator is the key observation: the empirical estimate has no finite variance and is not reliable.

File: scipy/stats/tests/test_discrete_basic.py
Comment (dschmitz89) [reply scipy/stats/tests/test_discrete_basic.py:95]: For me, with $a>8$, all tests pass except the kurtosis with (5, 9.3, 1). Also with other parameters I cannot get the kurtosis test to pass.

**Comment by dschmitz89** [comment]
@chrisb83 : I think I addressed all your remarks. Let me know if there is anything more to check.

File: scipy/stats/tests/test_discrete_basic.py
Comment (chrisb83) [reply scipy/stats/tests/test_discrete_basic.py:94]: strange, I did not run the test itself, I just copied the functionality from `common_tests.py` which should check the following:

```
dist = betanbinom(5, 9.3, 1)
m, v, s, k = dist.stats('mvsk')

if np.isfinite(k):
    m4e = dist.expect(lambda x: np.power(x-m, 4))
    assert_allclose(m4e, (k + 3.) * np.power(v, 2),
                        atol=1e-5, rtol=1e-5,
                        err_msg='kurtosis')
elif not np.isposinf(k):
    assert np.isnan(k)
```

and check does not fail when I run it

Anyways, we have enough checks around kurtosis now

**Review by chrisb83 — APPROVED** [review]
The current version looks good to me

**Comment by dschmitz89** [comment]
Gentle ping here: can this be merged?

**Comment by jachymb** [comment]
Hey, I appreciate the work done! I unfortunately didn't have the energy to follow the deveoplment, but I'm glad this is likely getting added!

**Comment by dschmitz89** [comment]
I added the milestone; this PR has one core dev approval and no requested changes anymore by anyone else.

**Comment by chrisb83** [comment]
Thanks for the reminder, Daniel! I am happy to merge this PR on Friday if there is no further feedback.

**Comment by chrisb83** [comment]
Thanks Daniel, I have just merged this PR

**Pull request merged**
